riscv-sbi-prop.md

```diff
@@ -0,0 +1,290 @@
+| sbi_hart_down             | Hart PM           | 0x01 000001  |-          |
+| sbi_hart_suspend          | Hart PM           | 0x01 000002  |-          |
+| sbi_hart_state            | Hart PM           | 0x01 000004  |-          |
+
+#### Function Description
+
+This section describes every newly introduced(in v0.2) function in details.
+Please refer to [1] for any v0.1 functions.
+
+```
+u32 sbi_get_version(void):
+```
+Returns the current SBI version implemented by the firmware.
+version: uint32: Bits[31:16] Major Version
+             Bits[15:0] Minor Version
+
+The existing SBI version can be 0.1. The proposed version will be at 0.2
+A different major version may indicate possible incompatible functions.
+A different minor version must be compatible with each other even if
+they have a higher number of features.
+
+```
+u32 sbi_get_function_mask(u32 ftype)
+```
+Given a function set type, it returns a bitmask of all functions IDs that are
+implemented for that function set.
+
+```
+u32 sbi_get_function_count(u32 ftype, u32 start_fid, unsigned long count):
+```
+
+This is a **reserved** function that should only be used if overflow bit in SBI
+function ID is set.
+
+Accepts a start_Function_id as an argument and returns if start_Function_id to
+(start_Function_id + count - 1) are supported or not.
+The Function numbering scheme is described in section 3.
+
+A count can help in minimizing the number of the M-mode traps by checking a range
+of SBI functions together.
+
+```
+int sbi_hart_up(unsigned long hartid, unsigned long start, unsigned
+long priv)
+```
+Brings up "hartid" either during initial boot or after a sbi_hart_down
+SBI call.
+
+"start" points to a runtime-specified address where a hart can enter
+into supervisor mode. This must be a physical address.
+
+"priv" is a private data that caller can use to pass information about
+execution context.
+
+Return the appropriate SBI error code.
+
+```
+int sbi_hart_suspend(u32 state, unsigned long resume_entry, unsigned
+long priv)
+```
+Suspends the calling hart to a particular power state. Suspended hart
+will automatically wake-up based on some wakeup events at resume_entry
+physical address.
+
+"priv" is a private data that caller can use to pass information about
+execution context. The SBI implementation must save a copy so that
+caller can reuse while restoring hart from suspend.
+
+Return the appropriate SBI error code.
+
+```
+int sbi_hart_down()
+```
+It powers off the hart and will be used in cpu-hotplug.
+Only individual hart can remove itself from supervisor mode. It can be
+moved to normal state only by sbi_hart_up function.
+
+Return the appropriate SBI error code.
+
+```
+u32 sbi_hart_state(unsigned long hartid)
+```
+Returns the RISCV_POWER_STATE for a specific hartid. This will help make
+kexec like functionality more robust.
+
+```
+void sbi_system_reset(u32 reset_type)
+```
+Reset the entire system.
+
+## Return error code Table:
```

Comment:
don't do this. 

Error numbers should have high entropy, and not tie themselves to the
11 bits of the Linux kernel. 0, -1, and 1 are particularly bad
choices, as they can appear by mistake in many cases. Constants should
also follow this rule. 0-based enums are to be avoided. In coreboot,
we use values like 'DATA' and 'CODE', rather than 0 or 1, as
constants, and this makes for easy debugging and fault analysis, as
well as avoiding mistakes from using default uninitialized values like
0 or -1.

Debugging with just error numbers is frustrating and error prone. It
should be possible to read an error string after a call. The error
string should be stored in per-hart data areas. We accept that this is
an extra call, since errors should be rare, and the extra overhead is
acceptable. We put error strings in Akaros 5 years ago and thank ourselves
daily for that decision. As a 40 year firmware writer, I can tell you that error
strings in firmware are easy.

Reply:
The return value from individual SBI calls still in discussion as I described in calling convention section. That's why I did not make any changes to it. My personal preference was to return a struct from discussion happened at https://github.com/riscv/riscv-sbi-doc/pull/9

Having error strings are definitely helpful. But that will lead to a hart local state. Since all of us want SBI implementation to be minimalistic, is it worth the extra management ?

Reply:
I can tell you from experience that error strings, vs. just an errno, is a game changer. 

I don't see how, with the complexity of what you are proposing here, that you can avoid per-hart state in M mode. If you're going to do that, then you might as well have errstr.